src/lib/idmap/sss_idmap.c

```diff
@@ -1078,6 +1170,180 @@ add_dom_for_sid(struct sss_idmap_ctx *ctx,
     return err;
 }
 
+enum idmap_error_code sss_idmap_offset_identity(void *pvt, uint32_t range_size,
+                                                const char *input,
+                                                long long *offset)
+{
+    long long out;
+    char *endptr;
+
+    if (input == NULL || offset == NULL) {
+        return IDMAP_ERROR;
+    }
+
+    errno = 0;
+    out = strtoull(input, &endptr, 10);
+    if (errno != 0 || out >= range_size || *endptr != '\0'
+                   || endptr == input) {
+        return IDMAP_ERROR;
+    }
+
+    *offset = out;
+
+    return IDMAP_SUCCESS;
+}
+
+enum idmap_error_code sss_idmap_rev_offset_identity(struct sss_idmap_ctx *ctx,
+                                                    void *pvt, uint32_t id,
+                                                    char **_out)
+{
+    char *out;
+    int len;
+    int ret;
+
+    len = snprintf(NULL, 0, "%"PRIu32, id);
+    if (len <= 0 || len > SID_STR_MAX_LEN) {
+        return IDMAP_ERROR;
+    }
+
+    out = ctx->alloc_func(len + 1, ctx->alloc_pvt);
+    if (out == NULL) {
+        return IDMAP_OUT_OF_MEMORY;
+    }
+
+    ret = snprintf(out, len + 1, "%"PRIu32, id);
+    if (ret != len) {
+        ctx->free_func(out, ctx->alloc_pvt);
+        return IDMAP_ERROR;
+    }
+
+    *_out = out;
+    return IDMAP_SUCCESS;
+}
+
+struct sss_idmap_offset_murmurhash3_data offset_murmurhash3_data_default = { 0xdeadbeef };
+
+enum idmap_error_code sss_idmap_offset_murmurhash3(void *pvt,
+                                                   uint32_t range_size,
+                                                   const char *input,
+                                                   long long *offset)
+{
+    struct sss_idmap_offset_murmurhash3_data *offset_murmurhash3_data;
+    long long out;
+
+    if (input == NULL || offset == NULL) {
+        return IDMAP_ERROR;
+    }
+
+    if (pvt != NULL) {
+        offset_murmurhash3_data = (struct sss_idmap_offset_murmurhash3_data *) pvt;
+    } else {
+        offset_murmurhash3_data = &offset_murmurhash3_data_default;
+    }
+
+    out = murmurhash3(input, strlen(input), offset_murmurhash3_data->seed);
+
+    out %= range_size;
+
+    *offset = out;
+
+    return IDMAP_SUCCESS;
+}
+
+enum idmap_error_code sss_idmap_gen_to_unix(struct sss_idmap_ctx *ctx,
+                                            const char *domain_id,
+                                            const char *input,
+                                            uint32_t *_id)
+{
+    struct idmap_domain_info *idmap_domain_info;
+    struct idmap_domain_info *matched_dom = NULL;
+    long long offset;
+    uint32_t range_size;
+    enum idmap_error_code err;
+    idmap_offset_func *offset_func = sss_idmap_offset_murmurhash3;
+    void *offset_func_pvt = NULL;
```

Comment:
`&offset_murmurhash3_data_default`?

Reply:
This would be possible but I would prefer that only the offset function itself knows its defaults so that e.g. we can move it into a separate file and do not have to make `offset_murmurhash3_data_default` public.